http/src/v0/pin.rs

```diff
@@ -1,107 +1,98 @@
-}
-
-async fn list_inner<T: IpfsTypes>(_ipfs: Ipfs<T>) -> Result<impl Reply, Rejection> {
-    // interestingly conformance tests call this with `paths=cid&stream=true&arg=cid`
-    // this needs to be a stream of the listing
-    Err::<&'static str, _>(crate::v0::NotImplemented.into())
-}
-
-pub fn rm<T: IpfsTypes>(
-    ipfs: &Ipfs<T>,
-) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone {
-    with_ipfs(ipfs).and_then(rm_inner)
+    with_ipfs(ipfs)
+        .and(add::add_request())
+        .and_then(add::add_inner)
 }
 
-async fn rm_inner<T: IpfsTypes>(_ipfs: Ipfs<T>) -> Result<impl Reply, Rejection> {
-    Err::<&'static str, _>(crate::v0::NotImplemented.into())
+#[derive(Debug)]
+struct ListRequest {
+    // FIXME: should be Vec<IpfsPath>
+    arg: Vec<Cid>,
+    filter: PinFilter,
+    // FIXME: not sure if this is used
+    quiet: bool,
+    // FIXME copypaste
+    stream: bool,
+    timeout: Option<humantime::Duration>,
 }
 
-impl<'a> TryFrom<&'a str> for AddRequest {
+impl<'a> TryFrom<&'a str> for ListRequest {
     type Error = ParseError<'a>;
 
     fn try_from(q: &'a str) -> Result<Self, Self::Error> {
         use ParseError::*;
 
+        // TODO: similar arg with duplicates question as always
+
+        let parse = url::form_urlencoded::parse(q.as_bytes());
         let mut args = Vec::new();
-        let mut recursive = None;
-        let mut progress = None;
+        let mut filter = None;
+        let mut quiet = None;
+        let mut stream = None;
+        let mut timeout = None;
 
-        for (key, value) in url::form_urlencoded::parse(q.as_bytes()) {
+        for (key, value) in parse {
             let target = match &*key {
                 "arg" => {
-                    args.push(Cid::try_from(&*value).map_err(|e| InvalidCid("arg".into(), e))?);
+                    // FIXME: this should be IpfsPath
+                    args.push(
+                        Cid::try_from(&*value)
+                            .map_err(|e| ParseError::InvalidCid("arg".into(), e))?,
+                    );
                     continue;
                 }
-                "recursive" => &mut recursive,
-                "progress" => &mut progress,
+                "type" => {
+                    if filter.is_none() {
+                        // not parsing this the whole way as there might be hope to have this
+                        // function removed in the future.
+                        filter = Some(match &*value {
+                            "all" => PinFilter::All,
+                            "direct" => PinFilter::Direct,
+                            "recursive" => PinFilter::Recursive,
+                            "indirect" => PinFilter::Indirect,
+                            other => {
+                                return Err(ParseError::InvalidValue(
+                                    "type".into(),
+                                    other.to_owned().into(),
+                                ))
```

Comment:
nit: having an `impl FromStr for PinFilter` would make this more concise

Reply:
Committed not pushed.